File: e2b/envd/client/filesystem.go
package client

import (
	"context"
	"fmt"
	"net/http"
	"time"

	"connectrpc.com/connect"
	"github.com/openkruise/agents-api/e2b/envd/filesystem"
	"github.com/openkruise/agents-api/e2b/envd/filesystem/filesystemconnect"
)

// EntryType represents the type of a filesystem entry.
type EntryType string

const (
	EntryTypeFile    EntryType = "file"
	EntryTypeDir     EntryType = "directory"
	EntryTypeSymlink EntryType = "symlink"
)

// EntryInfo represents information about a filesystem entry.
type EntryInfo struct {
	Name          string
	Type          EntryType
	Path          string
	Size          int64
	Mode          uint32
	Permissions   string
	Owner         string
	Group         string
	ModifiedTime  time.Time
	SymlinkTarget *string
}

// Filesystem provides filesystem operations in the sandbox over the official
// envd Filesystem gRPC service.
type Filesystem struct {
	Rpc     filesystemconnect.FilesystemClient
	headers map[string]string
}

// NewFilesystem creates a new Filesystem instance backed by the envd gRPC client.
func NewFilesystem(rpc filesystemconnect.FilesystemClient, headers map[string]string) *Filesystem {
	return &Filesystem{
		Rpc:     rpc,
		headers: headers,
	}
}

// List lists entries in a directory.
func (f *Filesystem) List(ctx context.Context, path string, depth ...int32) ([]EntryInfo, error) {
	d := uint32(1)
	if len(depth) > 0 && depth[0] >= 1 {
		d = uint32(depth[0])
	}

	req := connect.NewRequest(&filesystem.ListDirRequest{
		Path:  path,
		Depth: d,
	})
	f.setRPCHeaders(req)

	resp, err := f.Rpc.ListDir(ctx, req)
	if err != nil {
		return nil, fmt.Errorf("failed to list directory: %w", err)
	}

	entries := make([]EntryInfo, 0, len(resp.Msg.Entries))
	for _, entry := range resp.Msg.Entries {
		entryType := mapFileType(entry.Type)
		if entryType == "" {
			continue
		}
		entries = append(entries, convertEntryInfo(entry))
	}

	return entries, nil
}

// Exists checks if a file or directory exists.
func (f *Filesystem) Exists(ctx context.Context, path string) (bool, error) {
	req := connect.NewRequest(&filesystem.StatRequest{Path: path})
	f.setRPCHeaders(req)

	_, err := f.Rpc.Stat(ctx, req)
	if err != nil {
		if connectErr, ok := err.(*connect.Error); ok && connectErr.Code() == connect.CodeNotFound {
			return false, nil
		}
		return false, fmt.Errorf("failed to check existence: %w", err)
	}

	return true, nil
}

// GetInfo returns information about a file or directory.
func (f *Filesystem) GetInfo(ctx context.Context, path string) (*EntryInfo, error) {
	req := connect.NewRequest(&filesystem.StatRequest{Path: path})
	f.setRPCHeaders(req)

	resp, err := f.Rpc.Stat(ctx, req)
	if err != nil {
		return nil, fmt.Errorf("failed to get file info: %w", err)
	}

	info := convertEntryInfo(resp.Msg.Entry)
	return &info, nil
}

// Remove removes a file or directory.
func (f *Filesystem) Remove(ctx context.Context, path string) error {
	req := connect.NewRequest(&filesystem.RemoveRequest{Path: path})
	f.setRPCHeaders(req)

	_, err := f.Rpc.Remove(ctx, req)
	if err != nil {
		return fmt.Errorf("failed to remove: %w", err)
	}

	return nil
}

// Rename renames/moves a file or directory.
func (f *Filesystem) Rename(ctx context.Context, oldPath, newPath string) (*EntryInfo, error) {
	req := connect.NewRequest(&filesystem.MoveRequest{
		Source:      oldPath,
		Destination: newPath,
	})
	f.setRPCHeaders(req)

	resp, err := f.Rpc.Move(ctx, req)
	if err != nil {
		return nil, fmt.Errorf("failed to rename: %w", err)
	}

	info := convertEntryInfo(resp.Msg.Entry)
	return &info, nil
}

// MakeDir creates a new directory (and all parent directories if needed).
func (f *Filesystem) MakeDir(ctx context.Context, path string) (bool, error) {
	req := connect.NewRequest(&filesystem.MakeDirRequest{Path: path})
	f.setRPCHeaders(req)

	_, err := f.Rpc.MakeDir(ctx, req)
	if err != nil {
		if connectErr, ok := err.(*connect.Error); ok && connectErr.Code() == connect.CodeAlreadyExists {
			return false, nil
		}
		return false, fmt.Errorf("failed to make directory: %w", err)
	}

	return true, nil
}

func mapFileType(ft filesystem.FileType) EntryType {
	switch ft {
	case filesystem.FileType_FILE_TYPE_FILE:
		return EntryTypeFile
	case filesystem.FileType_FILE_TYPE_DIRECTORY:
		return EntryTypeDir
	case filesystem.FileType_FILE_TYPE_SYMLINK:
		return EntryTypeSymlink
	default:
		return ""
	}
}

func convertEntryInfo(entry *filesystem.EntryInfo) EntryInfo {
	info := EntryInfo{
		Name:        entry.Name,
		Type:        mapFileType(entry.Type),
		Path:        entry.Path,
		Size:        entry.Size,
		Mode:        entry.Mode,
		Permissions: entry.Permissions,
		Owner:       entry.Owner,
		Group:       entry.Group,
	}

	if entry.ModifiedTime != nil {
		info.ModifiedTime = entry.ModifiedTime.AsTime()
	}

	if entry.SymlinkTarget != nil {
		target := *entry.SymlinkTarget
		info.SymlinkTarget = &target
	}

	return info
}

func (f *Filesystem) setRPCHeaders(req interface{ Header() http.Header }) {
	for k, v := range f.headers {
		req.Header().Set(k, v)
	}
}

File: k8s/java/src/main/java/io/openkruise/agents/client/e2b/sandbox/filesystem/Filesystem.java
package io.openkruise.agents.client.e2b.sandbox.filesystem;

import com.google.protobuf.util.JsonFormat;
import io.grpc.ManagedChannel;
import io.grpc.StatusRuntimeException;
import io.openkruise.agents.client.e2b.api.models.Sandbox;
import io.openkruise.agents.client.e2b.config.ConnectionConfig;
import io.openkruise.agents.client.e2b.envd.filesystem.*;
import io.openkruise.agents.client.e2b.exceptions.SandboxException;
import io.openkruise.agents.client.e2b.utils.RpcUtils;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

import java.io.IOException;
import java.io.Reader;
import java.net.HttpURLConnection;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;

public class Filesystem {
    public static final String SERVICE_NAME = "filesystem.Filesystem";
    private static final JsonFormat.Parser PROTO_PARSER = JsonFormat.parser().ignoringUnknownFields();
    private final Sandbox sandbox;
    private final FilesystemGrpc.FilesystemBlockingStub blockingStub;
    private final ConnectionConfig config;
    private final OkHttpClient httpClient;

    public Filesystem(Sandbox sandbox, ManagedChannel channel, ConnectionConfig config) {
        this.sandbox = Objects.requireNonNull(sandbox, "sandbox cannot be null");
        this.blockingStub = FilesystemGrpc.newBlockingStub(channel).withMaxInboundMessageSize(32 * 1024 * 1024)
            .withMaxOutboundMessageSize(32 * 1024 * 1024);
        this.config = Objects.requireNonNull(config, "ConnectionConfig cannot be null");

        this.httpClient = new OkHttpClient.Builder().connectTimeout(config.getRequestTimeoutMs(), TimeUnit.MILLISECONDS)
            .readTimeout(config.getRequestTimeoutMs(), TimeUnit.MILLISECONDS).build();
    }

    /**
     * List entries in a directory.
     *
     * @param path path to the directory
     * @return list of entries in the directory
     */
    public List<EntryInfo> listDir(String path) {
        if (path == null || path.trim().isEmpty()) {
            throw new IllegalArgumentException("Path cannot be null or empty");
        }
        return listDir(path, 1);
    }

    /**
     * List entries in a directory with specified depth.
     *
     * @param path  path to the directory
     * @param depth depth of the directory to list
     * @return list of entries in the directory
     */
    public List<EntryInfo> listDir(String path, int depth) {
        if (path == null || path.trim().isEmpty()) {
            throw new IllegalArgumentException("Path cannot be null or empty");
        }
        if (depth < 1) {
            throw new IllegalArgumentException("Depth must be at least 1");
        }

        try {
            ListDirRequest params = ListDirRequest.newBuilder().setPath(path).setDepth(depth).build();

            Request request = config.buildHttpRequest(SERVICE_NAME, "ListDir", params, sandbox.getSandboxID());

            // noinspection WithSSRFCheckingInspection
            Response response = httpClient.newCall(request).execute();
            if (!response.isSuccessful()) {
                throw new IOException("ListDir HTTP request failed: " + response.code() + " " + response.message());
            }

            Reader reader = response.body().charStream();
            // Parse the JSON response back to protobuf
            ListDirResponse.Builder builder = ListDirResponse.newBuilder();
            PROTO_PARSER.merge(reader, builder);
            ListDirResponse listDirResponse = builder.build();

            List<EntryInfo> entries = new ArrayList<>(listDirResponse.getEntriesCount());
            for (io.openkruise.agents.client.e2b.envd.filesystem.EntryInfo entry :
                listDirResponse.getEntriesList()) {
                entries.add(toEntryInfo(entry));
            }
            return entries;
        } catch (Exception e) {
            throw new RuntimeException("Failed to listDir", e);
        }
    }

    /**
     * Create a new directory and all directories along the way if needed on the specified path.
     *
     * @param path path to a new directory
     * @return true if the directory was created, false if it already exists
     */
    public boolean makeDir(String path) {
        try {
            MakeDirRequest params = MakeDirRequest.newBuilder()
                .setPath(path)
                .build();

            Request request = config.buildHttpRequest(SERVICE_NAME, "MakeDir", params, sandbox.getSandboxID());

            // noinspection WithSSRFCheckingInspection
            Response response = httpClient.newCall(request).execute();
            if (!response.isSuccessful()) {
                throw new IOException("MakeDir HTTP request failed: " + response.code() + " " + response.message());
            }
            return response.isSuccessful();
        } catch (Exception e) {
            throw new RuntimeException("Failed to makeDir", e);
        }
    }

    /**
     * Rename a file or directory.
     *
     * @param oldPath path to the file or directory to rename
     * @param newPath new path for the file or directory
     * @return information about renamed file or directory
     */
    public boolean move(String oldPath, String newPath) {
        try {
            MoveRequest params = MoveRequest.newBuilder()
                .setSource(oldPath)
                .setDestination(newPath)
                .build();

            Request request = config.buildHttpRequest(SERVICE_NAME, "Move", params, sandbox.getSandboxID());

            // noinspection WithSSRFCheckingInspection
            Response response = httpClient.newCall(request).execute();
            if (!response.isSuccessful()) {
                throw new IOException("Move HTTP request failed: " + response.code() + " " + response.message());
            }

            return response.isSuccessful();
        } catch (Exception e) {
            throw new RuntimeException("Failed to move", e);
        }
    }

    /**
     * Remove a file or directory.
     *
     * @param path path to a file or directory
     */
    public boolean remove(String path) {
        try {
            RemoveRequest params = RemoveRequest.newBuilder()
                .setPath(path)
                .build();

            Request request = config.buildHttpRequest(SERVICE_NAME, "Remove", params, sandbox.getSandboxID());

            // noinspection WithSSRFCheckingInspection
            Response response = httpClient.newCall(request).execute();
            if (!response.isSuccessful()) {
                throw new IOException("Remove HTTP request failed: " + response.code() + " " + response.message());
            }
            return response.isSuccessful();
        } catch (Exception e) {
            throw new RuntimeException("Failed to remove", e);
        }
    }

    /**
     * Check if a file or a directory exists.
     *
     * @param path path to a file or a directory
     * @return true if the file or directory exists, false otherwise
     */
    public boolean exists(String path) {
        try {
            StatRequest params = StatRequest.newBuilder()
                .setPath(path)
                .build();

            Request request = config.buildHttpRequest(SERVICE_NAME, "Stat", params, sandbox.getSandboxID());

            // noinspection WithSSRFCheckingInspection
            Response response = httpClient.newCall(request).execute();
            if (response.code() == HttpURLConnection.HTTP_NOT_FOUND) {
                return false;
            }
            if (!response.isSuccessful()) {
                throw new IOException("Stat HTTP request failed: " + response.code() + " " + response.message());
            }
            return response.isSuccessful();
        } catch (Exception e) {
            throw new RuntimeException("Failed to exists", e);
        }
    }

    /**
     * Get information about a file or directory.
     *
     * @param path path to a file or directory
     * @return information about the file or directory
     */
    public EntryInfo getInfo(String path) {
        try {
            StatRequest params = StatRequest.newBuilder()
                .setPath(path)
                .build();

            Request request = config.buildHttpRequest(SERVICE_NAME, "Stat", params, sandbox.getSandboxID());

            // noinspection WithSSRFCheckingInspection
            Response response = httpClient.newCall(request).execute();
            if (!response.isSuccessful()) {
                throw new IOException("Stat HTTP request failed: " + response.code() + " " + response.message());
            }

            Reader reader = response.body().charStream();
            // Parse the JSON response back to protobuf
            StatResponse.Builder builder = StatResponse.newBuilder();
            PROTO_PARSER.merge(reader, builder);
            StatResponse statResponse = builder.build();

            return toEntryInfo(statResponse.getEntry());
        } catch (Exception e) {
            throw new RuntimeException("Failed to Remove", e);
        }
    }

    /**
     * Start watching a directory for filesystem events.
     *
     * @param path    path to directory to watch
     * @param onEvent callback to call when an event in the directory occurs
     * @return WatchHandle object for stopping watching directory
     */
    public WatchHandle watchDir(String path, Consumer<WatchHandle.FilesystemEvent> onEvent) {
        return watchDir(path, false, onEvent);
    }

    /**
     * Start watching a directory for filesystem events.
     *
     * @param path      path to directory to watch
     * @param recursive whether to watch subdirectories recursively
     * @param onEvent   callback to call when an event in the directory occurs
     * @return WatchHandle object for stopping watching directory
     */
    public WatchHandle watchDir(String path, boolean recursive, Consumer<WatchHandle.FilesystemEvent> onEvent) {
        WatchDirRequest request = WatchDirRequest.newBuilder().setPath(path).setRecursive(recursive).build();

        try {
            // This is a blocking stub, so we'll need to implement the streaming logic differently
            // For now, we'll create a streaming stub for this specific operation
            FilesystemGrpc.FilesystemStub streamingStub = FilesystemGrpc.newStub(blockingStub.getChannel());

            // Create a stream observer to handle responses
            io.grpc.stub.StreamObserver<WatchDirResponse> responseObserver = new WatchDirResponseObserver(onEvent);

            streamingStub.withDeadlineAfter(config.getRequestTimeoutMs(), TimeUnit.MILLISECONDS).watchDir(request,
                responseObserver);

            return new WatchHandle(responseObserver);
        } catch (StatusRuntimeException e) {
            throw RpcUtils.handleRpcException(e);
        }
    }

    private EntryInfo toEntryInfo(io.openkruise.agents.client.e2b.envd.filesystem.EntryInfo entry) {
        return new EntryInfo(entry.getName(), mapFileType(entry.getType()), entry.getPath(), entry.getSize(),
            entry.getMode(), entry.getPermissions(), entry.getOwner(), entry.getGroup(),
            entry.hasModifiedTime() ? entry.getModifiedTime() : null,
            entry.hasSymlinkTarget() ? entry.getSymlinkTarget() : null);
    }

    private FileType mapFileType(io.openkruise.agents.client.e2b.envd.filesystem.FileType fileType) {
        switch (fileType) {
            case FILE_TYPE_DIRECTORY:
                return FileType.DIR;
            case FILE_TYPE_FILE:
                return FileType.FILE;
            case FILE_TYPE_SYMLINK:
                return FileType.SYMLINK;
            default:
                throw new SandboxException("Unknown file type: " + fileType);
        }
    }

    public enum FileType {
        FILE,
        DIR,
        SYMLINK
    }

    public static class EntryInfo {
        private final String name;
        private final FileType type;
        private final String path;
        private final long size;
        private final int mode;
        private final String permissions;
        private final String owner;
        private final String group;
        private final com.google.protobuf.Timestamp modifiedTime; // 使用正确的Timestamp类型
        private final String symlinkTarget;

        public EntryInfo(String name, FileType type, String path, long size, int mode, String permissions, String owner,
            String group, com.google.protobuf.Timestamp modifiedTime, String symlinkTarget) {
            this.name = name;
            this.type = type;
            this.path = path;
            this.size = size;
            this.mode = mode;
            this.permissions = permissions;
            this.owner = owner;
            this.group = group;
            this.modifiedTime = modifiedTime;
            this.symlinkTarget = symlinkTarget;
        }

        // Getters
        public String getName() {return name;}

        public FileType getType() {return type;}

        public String getPath() {return path;}

        public long getSize() {return size;}

        public int getMode() {return mode;}

        public String getPermissions() {return permissions;}

        public String getOwner() {return owner;}

        public String getGroup() {return group;}

        public com.google.protobuf.Timestamp getModifiedTime() {return modifiedTime;} // 更准确的类型

        public String getSymlinkTarget() {return symlinkTarget;}

        @Override
        public String toString() {
            return String.format("EntryInfo{name='%s', type=%s, path='%s', size=%d}", name, type, path, size);
        }
    }
}
